fastcache.go

```diff
@@ -121,6 +121,40 @@ func (c *Cache) UpdateStats(s *Stats) {
 	}
 }
 
+// VisitAllEntries calls f for all the cache entries.
+//
+// The function returns immediately if f returns non-nil error.
+// It returns the given error.
+//
+// f cannot hold pointers to k and v contents after returning.
+func (c *Cache) VisitAllEntries(f func(k, v []byte) error) error {
+	for _, b := range c.buckets {
+		b.mu.RLock()
+		for _, idx := range b.m {
+			idx &= (1 << bucketSizeBits) - 1
+			chunkIdx := idx / chunkSize
+			chunk := b.chunks[chunkIdx]
+
+			kvLenBuf := chunk[idx : idx+4]
+			keyLen := (uint64(kvLenBuf[0]) << 8) | uint64(kvLenBuf[1])
+			valLen := (uint64(kvLenBuf[2]) << 8) | uint64(kvLenBuf[3])
+
+			idx += 4
+			key := chunk[idx : idx+keyLen]
+
+			idx += keyLen
+			value := chunk[idx : idx+valLen]
+
+			if err := f(key, value); err != nil {
```

Comment:
If we mention these possible situations in the docs, do you think users will still ignore them and they will end up in undesired states?

Reply:
> If we mention these possible situations in the docs, do you think users will still ignore them and they will end up in undesired states?

I'm absolutely sure :) There are the following choices:
- to document all the restrictions and caveats in the hope users read and adhere the docs
- to harden the implementation against users who don't read docs. I.e. do not hold the lock during `f` call and copy `key`, `value` args passed to `f`. This would slow down the implementation
- to fork `fastcache` and implement this in the fork

I'm leaning to the last solution. Let's leave the PR open for a while.

Reply:
Could we push on a queue all operations (get, set, del...) while VisitAllEntries is working? And when it finishes, we run all the operations (in the same order, of course).